lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,306 @@
+
+  /// Creates a new [FixedDateTimeFormatter] with the provided [pattern].
+  ///
+  /// The [pattern] interprets the characters mentioned in
+  /// [FixedDateTimeFormatter] to represent fields of a `DateTime` value. Other
+  /// characters are not special. If [isUtc] is set to false, the DateTime is
+  /// constructed with respect to the local timezone.
+  ///
+  /// There must at most be one sequence of each special character to ensure a
+  /// single source of truth when constructing the [DateTime], so a pattern of
+  /// `"CCCC-MM-DD, CC"` is invalid, because it has two separate `C` sequences.
+  FixedDateTimeFormatter(this.pattern, {this.isUtc = true}) {
+    int? current;
+    var start = 0;
+    var characters = pattern.codeUnits;
+    for (var i = 0; i < characters.length; i++) {
+      var char = characters[i];
+      if (current != char) {
+        _blocks.saveBlock(current, start, i);
+        if (_validFormatCharacters.contains(char)) {
+          var hasSeenBefore = _blocks.formatCharacters.indexOf(char);
+          if (hasSeenBefore > -1) {
+            throw FormatException(
+                "Pattern contains more than one '$char' block.\n"
+                "Previous occurrence at index ${_blocks.starts[hasSeenBefore]}",
+                pattern,
+                i);
+          } else {
+            start = i;
+            current = char;
+          }
+        } else {
+          current = null;
+        }
+      }
+    }
+    _blocks.saveBlock(current, start, pattern.length);
+  }
+
+  /// Convert [DateTime] to a [String] exactly as specified by the [pattern].
+  String encode(DateTime datetime) {
+    var buffer = StringBuffer();
+    for (var i = 0; i < _blocks.length; i++) {
+      var start = _blocks.starts[i];
+      var end = _blocks.ends[i];
+      var length = end - start;
+
+      var previousEnd = i > 0 ? _blocks.ends[i - 1] : 0;
+      if (previousEnd < start) {
+        buffer.write(pattern.substring(previousEnd, start));
+      }
+      var formatCharacter = _blocks.formatCharacters[i];
+      var number = _extractNumFromDateTime(
+        formatCharacter,
+        datetime,
+      ).toString();
+      if (number.length > length) {
+        if (formatCharacter == _fractionSecondCode) {
+          //Special case, as we want fractional seconds to be the leading digits
+          number = number.substring(length);
+        } else {
+          number = number.substring(number.length - length);
+        }
+      } else if (number.length < length) {
+        number = number.padLeft(length, '0');
+      }
+      buffer.write(number);
+    }
+    if (_blocks.length > 0) {
+      var lastEnd = _blocks.ends.last;
+      if (lastEnd < pattern.length) {
+        buffer.write(pattern.substring(lastEnd, pattern.length));
+      }
+    }
+    return buffer.toString();
+  }
+
+  int _extractNumFromDateTime(int? formatChar, DateTime dateTime) {
+    switch (formatChar) {
+      case _yearCode:
+        return dateTime.year;
```

Comment:
Consider whether this should return `dateTime.year.abs()`. See below.

Reply:
See my comment above :)